analytics/transform.go

```diff
@@ -0,0 +1,270 @@
+	}
+	return []KeyVal{KeyVal{key, value}}, nil
+}
+
+func parseInt(key string, value string) ([]KeyVal, error) {
+	if value == "" {
+		return nil, errors.Wrap(errors.New("Null string found"), fmt.Sprintf("Error parsing key %s", key))
+	}
+	intValue, err := strconv.Atoi(value)
+	if err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("Error parsing key '%s' to integer", key))
+	}
+	return []KeyVal{KeyVal{key, intValue}}, err
+}
+
+func parseBool(key string, value string) ([]KeyVal, error) {
+	if value == "" {
+		return nil, errors.Wrap(errors.New("Null string found"), fmt.Sprintf("Error parsing key %s", key))
+	}
+	boolValue, err := strconv.ParseBool(value)
+	if err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("Error parsing key '%s' to boolean", key))
+	}
+	return []KeyVal{KeyVal{key, boolValue}}, err
+}
+
+func parseDouble(key string, value string) ([]KeyVal, error) {
+	if value == "" {
+		return nil, errors.Wrap(errors.New("Null string found"), fmt.Sprintf("Error parsing key %s", key))
+	}
+	doubleValue, err := strconv.ParseFloat(value, 64)
+	if err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("Error parsing key '%s' to double", key))
+	}
+	return []KeyVal{KeyVal{key, doubleValue}}, err
+}
+
+func parseContexts(key string, value string) ([]KeyVal, error) {
+	if value == "" {
+		return nil, errors.Wrap(errors.New("Null string found"), fmt.Sprintf("Error parsing key %s", key))
+	}
+	return shredContexts(value)
+}
+
+func parseUnstruct(key string, value string) ([]KeyVal, error) {
+	if value == "" {
+		return nil, errors.Wrap(errors.New("Null string found"), fmt.Sprintf("Error parsing key %s", key))
+	}
+	return shredUnstruct(value)
+}
+
+func mapifyGoodEvent(event []string, knownFields [131]KeyFunctionPair, addGeolocationData bool) (map[string]interface{}, error) {
+	if len(event) != len(knownFields) {
+		return nil, errors.New("Cannot transform event - wrong number of fields")
+	} else {
+		output := make(map[string]interface{})
+		if addGeolocationData && event[latitudeIndex] != "" && event[longitudeIndex] != "" {
+			output["geo_location"] = event[latitudeIndex] + "," + event[longitudeIndex]
+		}
+		for index, value := range event {
+			// skip if empty
+			if event[index] != "" {
+				// apply function if not empty
+				kvPairs, err := knownFields[index].ParseFunction(knownFields[index].Key, value)
+				if err != nil {
+					return nil, err
+				}
+				// append all results
+				for _, pair := range kvPairs {
+					output[pair.Key] = pair.Value
+				}
+			}
+		}
+		return output, nil
+	}
+}
+
+// Design decision made: Separated functions used to add Geo fields.
+// These could be replaced by moving the functionality to the GetSubset functions - but it might be less intuitive.
+
+// ToMap transforms a valid tsv string Snowplow event to a Go map.
+func ToMap(event string) (map[string]interface{}, error) {
+	record := strings.Split(event, "\t")
```

Comment:
It'd be nice to figure out how to extract this out of the functions, perhaps into a higher level object or something, so that you only need to split the event TSV once then you can call `Get*` and `To*` without needing to recompute the splitting.
I don't think this is required for 0.1.0, as for most use cases I think you'll call one of the methods in particular and then move on but for usecases where you want to call it twice then you won't want to be calling `Split` multiple times. Maybe worth filing an issue for 0.2.0, I think the only reason I'd be tempted to do it for 0.1.0 is I anticipate this will cause an API change.

Reply:
I'm struggling a bit to conceptualise what you have in mind here. Perhaps we can chat about it whenever suits?

Reply:
I'm imagining something like:

```
var parsedEvent = parse(event) // This could be an interface/type to hide implementation details although ultimately is probably just the split string return.
parsedEvent.toJson();
parsedEvent.Get*();
```

Reply:
Ok I follow. I didn't realise you meant a change to the public interface rather than internals. 

I see what you mean now, will take a look. :)